### PR TITLE
feat: add Topsort API v2

### PR DESCRIPTION
**Pull request by roalonso**
Check in https://editor.swagger.io/

### CODE REVIEW

**Review by PabloReszczynski — APPROVED** [review]
lgtm

File: topsort-api-v2.yml [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+                        listings: '#/components/schemas/AuctionRequestSponsoredListings'
+                        bannerAds: '#/components/schemas/AuctionRequestBannerAds'
+        required: true
+      responses:
+        201:
+          description: >
+            The auction results.
+            The list of winners will contain at most `slots` entries per auction.
+            It may contain fewer or no entries at all if there aren't enough products with usable bids.
+            Bid become unusable if campaign budget is exhausted, the bid is disqualified to preserve spend pacing, etc.
```
Comment (sk-) [inline topsort-api-v2.yml:127]:
```suggestion
            Bids become unusable if campaign budget is exhausted, the bid is disqualified to preserve spend pacing, etc.
```

File: topsort-api-v2.yml [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+                      mapping:
+                        listings: '#/components/schemas/AuctionRequestSponsoredListings'
+                        bannerAds: '#/components/schemas/AuctionRequestBannerAds'
+        required: true
+      responses:
+        201:
+          description: >
+            The auction results.
+            The list of winners will contain at most `slots` entries per auction.
+            It may contain fewer or no entries at all if there aren't enough products with usable bids.
```
Comment (sk-) [inline topsort-api-v2.yml:126]: In the definition of usable there's no mention at all about being above the reserve price.

File: topsort-api-v2.yml [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+                  type: array
+                  items:
+                    oneOf:
+                      - $ref: '#/components/schemas/AuctionRequestSponsoredListings'
+                      - $ref: '#/components/schemas/AuctionRequestBannerAds'
+                    discriminator:
+                      propertyName: type
+                      mapping:
+                        listings: '#/components/schemas/AuctionRequestSponsoredListings'
+                        bannerAds: '#/components/schemas/AuctionRequestBannerAds'
```
Comment (sk-) [inline topsort-api-v2.yml:119]: It has always troubled me that bannerAds and videoAds have the `Ads` suffix, but not the `listings`.

Also `listings` could be `sponsoredProducts`

File: topsort-api-v2.yml [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+                          winnerType: product
+                          winnerId: e_PJbnN
+                          resolvedBidId: 3aaa25fe-60a0-45a3-ae65-39ec1b525eb4
+                          assetUrl: https://address.to/lhs-banner-image-for-e_PJbnN.png
+        401:
+          $ref: '#/components/responses/UnauthorizedError'
+        400:
+          $ref: '#/components/responses/BadRequest'
+
+  /events:
```
Comment (sk-) [inline topsort-api-v2.yml:159]: I wouldn't add the events endpoint yet. We need to reorganize that API, which has a lot of unused fields and it's not batchable.

File: topsort-api-v2.yml [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+          Topsort will run an auction for each batched auction request, for which products, brands and/or vendors' bids will compete against each other.
+          Provided fields are additive, in the sense that only bids that match all filtering fields are considered (e.g. in the example below, only bids for products in the c_yogurt category that match the "Noosa Peach" search string will participate in the acution).
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                auctions:
+                  type: array
+                  items:
```
Comment (sk-) [inline topsort-api-v2.yml:111]: What's the maximum number of auctions we are going to process in parallel?

File: topsort-api-v2.yml [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+          type: string
+          example: c_yogurt
+
+    SearchQuery:
+      type: string
+      description: The search string provided by a user.
+
+    GeoTargeting:
+      type: object
+      description: An object describing geographical information associated with this auction.
```
Comment (sk-) [inline topsort-api-v2.yml:247]: There must be a clarification indicating how that location is obtained, or how it's configured across their and our system.

Something like, this id is the same that you configured in the Toposort manager.

File: topsort-api-v2.yml [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+    Product:
+      required:
+        - productId
+      type: object
+      properties:
+        productId:
+          type: string
+          format: string
+          example: p_SA0238
+        quality:
```
Comment (sk-) [inline topsort-api-v2.yml:369]: Can we remove this, it's not used at the moment.

File: topsort-api-v2.yml [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+      description: Describes the intent of running a banner ads auction.
+      allOf:
+        - $ref: '#/components/schemas/AuctionRequest'
+        - type: object
+          properties:
+            slotId:
+              type: string
+            category:
+              $ref: '#/components/schemas/Category'
+            searchQuery:
```
Comment (sk-) [inline topsort-api-v2.yml:303]: Can we just use string instead her.

File: topsort-api-v2.yml [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+                properties:
+                  results:
+                    type: array
+                    items:
+                      $ref: '#/components/schemas/Auction'
+                example:
+                  results:
+                    - winners:
+                        - rank: 0
+                          winnerType: product
```
Comment (sk-) [inline topsort-api-v2.yml:141]: I'd remove the prefix `winner`, as you are already inside a winner object.

**Review by sk- — CHANGES_REQUESTED** [review]

**Comment by sk-** [comment]
Also, please generate the API using openapi generator in a couple of languages. I know that sometimes it doesn not work too well when using anonymous types.

**Pull request merged**